ECS-Agent image deletion stops if non-existant image encountered

<!--
For urgent operational issues, please contact AWS Support directly.
https://aws.amazon.com/premiumsupport/

Please provide the following information:
-->


### Summary
<!-- Please provide a brief outline of the issue -->
The ECS-Agent image deletion / clean-up stops with an error if it encounters a reference to an image that was removed manually via Docker. It processes no other candidate images following this error, nor does it ignore the absent image during future clean-up cycles.


### Description
<!-- Provide detailed information about this issue -->
The following command was issued on an EC2 container instance in response to low disk space:
`docker rmi $(docker images -q)`
This was intended to remove all images not currently associated with running or stopped containers.

However, the ECS-Agent now repeatedly looks for these absent images and will error out after not being able to locate the first one. Ex:
```
2019-02-19T20:08:34Z [INFO] Found 52 eligible images for deletion
2019-02-19T20:08:34Z [INFO] Removing Image: mycompany/goms-rest-staging:latest
2019-02-19T20:08:34Z [ERROR] Error removing Image mycompany/goms-rest-staging:latest - Error: No such image: mycompany/goms-rest-staging:latest
```
At this point, the clean-up stops, and no further image deletion is attempted. The result is a steadily increasing number of images that are not used by running containers.

<!-- Not required for feature requests -->
### Expected Behavior
Ideally, the ECS-Agent should recognize that the image is no longer available and continue to the next candidate image. In addition, it should remove the image from its tracking list such that it is not referenced again as a "candidate".

Another acceptable alternative would be to reset the agent's image tracking list such that it only tracked images related to managed containers on start-up

### Observed Behavior


### Environment Details
<!--
Examples:
* docker info
* curl http://localhost:51678/v1/metadata
* df -h
-->


### Supporting Log Snippets
<!--
Please have a look at https://github.com/awslabs/ecs-logs-collector
for the data we typically require for investigation

Please note that GitHub issues are public, remove sensitive data from logs before posting.
If you are not comfortable posting your logs here, please let us know and we
can provide an alternate method.
-->

--- a/agent/engine/docker_image_manager.go
+++ b/agent/engine/docker_image_manager.go
@@ -17,6 +17,7 @@ import (
 	"context"
 	"fmt"
 	"sort"
+	"strings"
 	"sync"
 	"time"
 
@@ -537,7 +538,7 @@ func (imageManager *dockerImageManager) deleteImage(ctx context.Context, imageID
 	seelog.Infof("Removing Image: %s", imageID)
 	err := imageManager.client.RemoveImage(ctx, imageID, dockerclient.RemoveImageTimeout)
 	if err != nil {
-		if err.Error() == imageNotFoundForDeletionError {
+		if strings.Contains(err.Error(), imageNotFoundForDeletionError) {
 			seelog.Errorf("Image already removed from the instance: %v", err)
 		} else {
 			seelog.Errorf("Error removing Image %v - %v", imageID, err)

--- a/agent/engine/docker_image_manager_test.go
+++ b/agent/engine/docker_image_manager_test.go
@@ -1064,7 +1064,9 @@ func TestDeleteImage(t *testing.T) {
 	}
 }
 
-func TestDeleteImageNotFoundError(t *testing.T) {
+// This test tests that we detect correctly in agent when the agent is trying to delete image that
+// does not exist for older version of docker. 
+func TestDeleteImageNotFoundOldDockerMessageError(t *testing.T) {
 	ctrl := gomock.NewController(t)
 	defer ctrl.Finish()
 	client := mock_dockerapi.NewMockDockerClient(ctrl)
@@ -1085,6 +1087,38 @@ func TestDeleteImageNotFoundError(t *testing.T) {
 	imageState, _ := imageManager.getImageState(imageInspected.ID)
 	client.EXPECT().RemoveImage(gomock.Any(), container.Image, dockerclient.RemoveImageTimeout).Return(
 		errors.New("no such image"))
+	ctx, cancel := context.WithCancel(context.TODO())
+	defer cancel()
+	imageManager.deleteImage(ctx, container.Image, imageState)
+	if len(imageState.Image.Names) != 0 {
+		t.Error("Error removing Image name from image state")
+	}
+	if len(imageManager.getAllImageStates()) != 0 {
+		t.Error("Error removing image state from image manager")
+	}
+}
+
+func TestDeleteImageNotFoundError(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+	client := mock_dockerapi.NewMockDockerClient(ctrl)
+	imageManager := &dockerImageManager{client: client, state: dockerstate.NewTaskEngineState()}
+	imageManager.SetSaver(statemanager.NewNoopStateManager())
+	container := &apicontainer.Container{
+		Name:  "testContainer",
+		Image: "testContainerImage",
+	}
+	imageInspected := &types.ImageInspect{
+		ID: "sha256:qwerty",
+	}
+	client.EXPECT().InspectImage(container.Image).Return(imageInspected, nil).AnyTimes()
+	err := imageManager.RecordContainerReference(container)
+	if err != nil {
+		t.Error("Error in adding container to an existing image state")
+	}
+	imageState, _ := imageManager.getImageState(imageInspected.ID)
+	client.EXPECT().RemoveImage(gomock.Any(), container.Image, dockerclient.RemoveImageTimeout).Return(
+		errors.New("no such image: " + container.Image))
 	ctx, cancel := context.WithCancel(context.TODO())
 	defer cancel()
 	imageManager.deleteImage(ctx, container.Image, imageState)

